package-res/resources/web/pentaho/type/ComplexChangeset.js

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+
+define([
+  "../lang/Base",
+  "../util/object",
+  "../util/error"
+], function(Base, O, error) {
+  "use strict";
+
+  /**
+   * @classDesc A 'ComplexChangeset' represents a collection of changes in a set of properties.
+   *
+   * ### `ComplexChangeset` Property Key
+   * A `ComplexChangeset` property key is the name of a complex's `property` that had its value changed.
+   *
+   * ### `ComplexChangeset` Change Object
+   * A `ComplexChangeset` `change` object describes the changes to be made to a `property`.
+   *
+   * @name ComplexChangeset
+   * @memberOf pentaho.lang
+   *
+   * @amd pentaho/lang/ComplexChangeset
+   * @class
+   * @extends pentaho.lang.Base
+   */
+  return Base.extend("pentaho.type.ComplexChangeset", /** @lends pentaho.type.ComplexChangeset#*/{
+
+    /**
+     * Creates a `ComplexChangeset` with a given owner.
+     *
+     * @constructor
+     *
+     * @param {!pentaho.type.Complex} owner - The complex where the change occurred.
+     */
+    constructor: function(owner) {
+      if(!owner) throw error.argRequired("owner");
+
+      this._owner = owner;
+      this._properties = {};
+    },
+
+    /**
+     * Gets the complex where the change occurred.
+     *
+     * @type !pentaho.type.ComplexChangeset
+     * @readonly
+     */
+    get owner() {
+      return this._owner;
+    },
+
+    /**
+     * Determines if the `ComplexChangeset` contains the specified property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {boolean} `true` if the property exists, `false` otherwise.
+     */
+    has: function(name) {
+      var pType = this.owner.type.get(name);
+      if(pType.isList) {
+        //does nothing for now
+        return false;
+      } else {
+        return O.hasOwn(this._properties, name);
+      }
+    },
+
+    /**
+     * Sets the value of a property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     * @param {any?} [valueSpec=null] A value specification.
+     */
+    set: function(name, valueSpec) {
+      if(!name) throw error.argRequired("property");
+
+      var pType = this.owner.type.get(name);
+      if(pType.isList) {
+        //TODO: does nothing for now
+      } else {
+        this._setValueChange(name, pType.toValue(valueSpec));
+      }
+    },
+
+    /**
+     * Gets the change object mapped to the specified property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {?pentaho.lang.ValueChange} An object containing the changes to be operated
+     * in the given property, or `null` if the property is not defined in this changeset.
+     */
+    get: function(name) {
+      if(!this.has(name)) return null;
+      var propertyName = typeof name === "string" ? name : this.owner.type.get(name).name;
```

Comment:
Currently,  the call `this.owner.type.get(name)` does not throw on undefined property names.